cmd/kube-proxy/app/server.go

```diff
@@ -252,7 +253,33 @@ func NewProxyServerDefault(config *ProxyServerConfig) (*ProxyServer, error) {
 		glog.V(2).Info("Using iptables Proxier.")
 		proxierIptables, err := iptables.NewProxier(iptInterface, execer, config.IptablesSyncPeriod, config.MasqueradeAll)
 		if err != nil {
-			glog.Fatalf("Unable to create proxier: %v", err)
+			iptables_err := err.Error()
+			iptables_err_str := ""
+			if strings.Contains(iptables_err, "iptables") {
+				//	Error is most probably caused by lack of iptables
+				// configuration
+				iptables_err_str = "Check if iptables has been enabled in the kernel. " +
```

Comment:
That's one of the very few that will trigger at init time (this error path)
and even then, only in userspace mode.  The iptables mode proxier (which
uses many more features) doesn't actually do any iptables work until later.

My point being this error is helpful IF it triggers, but I think you need
to do something like find all the places where we call iptables to install
some rules (Ensure*, Restore, etc) and print this at that time.  But only
do it once - keep a flag variable indicating that it has been logged
already.

What do you think of that?  More invasive but more likely to help someone.

On Tue, Nov 24, 2015 at 10:17 PM, asridharan <notifications@github.com>
wrote:

> In cmd/kube-proxy/app/server.go
> <https://github.com/kubernetes/kubernetes/pull/17272#discussion_r45832676>
> :
>
> > @@ -252,7 +253,33 @@ func NewProxyServerDefault(config *ProxyServerConfig) (*ProxyServer, error) {
> >  		glog.V(2).Info("Using iptables Proxier.")
> >  		proxierIptables, err := iptables.NewProxier(iptInterface, execer, config.IptablesSyncPeriod, config.MasqueradeAll)
> >  		if err != nil {
> > -			glog.Fatalf("Unable to create proxier: %v", err)
> > +			iptables_err := err.Error()
> > +			iptables_err_str := ""
> > +			if strings.Contains(iptables_err, "iptables") {
> > +				//	Error is most probably caused by lack of iptables
> > +				// configuration
> > +				iptables_err_str = "Check if iptables has been enabled in the kernel. " +
>
> I did a test by removing the xt_comment module to trigger this error.
> … <#1513d4a6e3de2390_>
> On Tue, Nov 24, 2015 at 9:54 PM Tim Hockin <notifications@github.com>
> wrote: In cmd/kube-proxy/app/server.go <#17272 (comment)
> <https://github.com/kubernetes/kubernetes/pull/17272#discussion_r45831825>>
> : > @@ -252,7 +253,33 @@ func NewProxyServerDefault(config
> *ProxyServerConfig) (*ProxyServer, error) { > glog.V(2).Info("Using
> iptables Proxier.") > proxierIptables, err :=
> iptables.NewProxier(iptInterface, execer, config.IptablesSyncPeriod,
> config.MasqueradeAll) > if err != nil { > - glog.Fatalf("Unable to create
> proxier: %v", err) > + iptables_err := err.Error() > + iptables_err_str :=
> "" > + if strings.Contains(iptables_err, "iptables") { > + // Error is most
> probably caused by lack of iptables > + // configuration > +
> iptables_err_str = "Check if iptables has been enabled in the kernel. " +
> My point was that I'm not sure this error will be triggered in the event of
> those modules being missing. If those modules are actually missing from the
> machine, I expect it to fail much later and much less obviously. Do we have
> evidence (e.g. manual tests) that this error actually fires? I don't see
> how it would. — Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/17272/files#r45831825>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17272/files#r45832676>.
>